components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java

```diff
@@ -492,6 +503,36 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx)
                 throw new IdentityOAuth2Exception(e.getMessage(), e);
             }
         }
+
+        ServiceProvider serviceProvider;
+        try {
+            serviceProvider = OAuth2ServiceComponentHolder.getApplicationMgtService().getServiceProviderByClientId(
+                    clientId, OAUTH2, tenantDomain);
+        } catch (IdentityApplicationManagementException e) {
+            throw new IdentityOAuth2Exception("Error occurred while retrieving OAuth2 application data for client id " +
+                    clientId, e);
+        }
+
+        String serviceProviderName = serviceProvider.getApplicationName();
+        // FIX for IDENTITY-4531
+        // if the service provider name is returned as "default" it means that a valid service provider for
+        // the given client ID was not found in the tenantDomain sent in the request
+        if (DEFAULT_SP_NAME.equals(serviceProviderName)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Valid Service provider not found for client id "+clientId +
+                        " and tenant domain "+tenantDomain);
+            }
+            return false;
+        }
+
+        if (!serviceProvider.isSaasApp() && !userTenantDomain.equals(tenantDomain)){
+            if(log.isDebugEnabled()) {
+                log.debug("Non-SaaS service provider tenant domain is not same as user tenant domain; " +
```

Comment:
The debug log was already in PasswordGrantHandler, I will change the log to reflect the idea clearly